### PR TITLE
Use correct Java version for JVM dependency repos.

### LINKED ISSUE #21882: PackageHub: infer correct JVM version for Maven packages
Currently, the JVM Packages external service type assumes that all packages are Java 8. We should infer the correct JVM version by reading the first bytes of the first classfile.

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, we used Java 8 for all JVM dependencies. This meant that a
Java 11 library would fail to index. This commit fixes that issue
by reading the classfiles of the given JVM dependency to infer the
actual version of the JVM that it's compatible with.

Fixes #21882.

### CODE REVIEW

**Review by Strum355 — APPROVED** [review]

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go [hunk 1]
```diff
@@ -263,6 +286,55 @@ func (s *JVMPackagesSyncer) commitJar(ctx context.Context, dependency reposource
...
+	if err != nil {
+		return "", err
+	}
+	majorVersion, err := strconv.Atoi(majorVersionString)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to convert string %s to int", majorVersion)
+	}
+
+	// Java 1.1 (aka "Java 1") has major version 45 and Java 8 has major
+	// version 52.  To go from the major version of Java version we subtract
```
Comment (efritz) [inline cmd/gitserver/server/vcs_syncer_jvm_packages.go:302]:
```suggestion
	// version 52. To go from the major version of Java version we subtract
```

I'm helping!

Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_jvm_packages.go]: Fixed.

Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_jvm_packages.go]: The double space came from my vim bindings that line wrap at 80 characters on `gq`

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go [hunk 2]
```diff
@@ -277,3 +349,63 @@ type lsifJavaJSON struct {
...
+		return "", errors.Wrapf(err, "failed to read jar file %s", byteCodeJarPath)
+	}
+
+	for _, zipEntry := range zipReader.File {
+		if strings.HasSuffix(zipEntry.Name, ".class") {
+			classFileReader, err := zipEntry.Open()
+			if err != nil {
+				return "", err
+			}
+			defer classFileReader.Close()
```
Comment (unknwon) [inline cmd/gitserver/server/vcs_syncer_jvm_packages.go:378]: The classic pitfall for having `defer` in a loop 😁 Just be noted all of those files won't be closed until the `classFileMajorVersion` returns. No problem in practice when the number of files is small though.

The easiest approach is to put the loop body in a separate (anonymous) function.

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go [hunk 3]
```diff
@@ -172,23 +178,34 @@ func (s *JVMPackagesSyncer) gitPushDependencyTag(ctx context.Context, bareGitDir
 	// Always clean up created temporary directories.
 	defer os.RemoveAll(tmpDirectory)
 
-	paths, err := coursier.FetchSources(ctx, s.Config, dependency)
+	sourceCodes, err := coursier.FetchSources(ctx, s.Config, dependency)
```
Comment (efritz) [inline cmd/gitserver/server/vcs_syncer_jvm_packages.go:181]: Can we call these `sourceCodePaths` and `byteCodePaths`? This was confusing to read first time through.

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go [hunk 4]
```diff
@@ -277,3 +349,63 @@ type lsifJavaJSON struct {
...
+	if err != nil {
+		return "", err
+	}
+	zipReader, err := zip.NewReader(file, stat.Size())
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to read jar file %s", byteCodeJarPath)
+	}
+
+	for _, zipEntry := range zipReader.File {
+		if strings.HasSuffix(zipEntry.Name, ".class") {
```
Comment (efritz) [inline cmd/gitserver/server/vcs_syncer_jvm_packages.go:373]: If the entire loop body is in an if consequence it's usually better to continue with the negation of that conditional instead (my personal preferred style).

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go [hunk 5]
```diff
@@ -263,6 +286,55 @@ func (s *JVMPackagesSyncer) commitJar(ctx context.Context, dependency reposource
...
+	if javaVersion <= 8 {
+		return 8
+	}
+	if javaVersion <= 11 {
+		return 11
+	}
+	// TODO: bump this up to 17 once Java 17 LTS has been released, see https://github.com/sourcegraph/lsif-java/issues/263
+	if javaVersion <= 16 {
+		return 16
+	}
```
Comment (efritz) [inline cmd/gitserver/server/vcs_syncer_jvm_packages.go:333]: This could be a pretty nifty loop

**Review by efritz — APPROVED** [review]

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go (hunk 3 above)
Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_jvm_packages.go]: Good point, I agree.

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go (hunk 5 above)
Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_jvm_packages.go:333]: I thought about doing something smarter but I feel like it's easier to read/maintain this way. The method only needs to be updated every 3 years with new LTS releases, and there's no way around that unless we do some kind of remote call to check the latest LTS release.

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go (hunk 4 above)
Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_jvm_packages.go]: Good point, done. I'm used to writing Scala for-comprehensions where you can write nested loops with flat indentation.

File: cmd/gitserver/server/vcs_syncer_jvm_packages.go (hunk 2 above)
Comment (olafurpg) [reply cmd/gitserver/server/vcs_syncer_jvm_packages.go]: Nice catch! I refactored the body of the for loop into a separate named function.

**Review by olafurpg — COMMENTED** [review]
Thank you for the review!

**Review by unknwon — APPROVED** [review]

**Comment by olafurpg** [comment]
Just wanted to confirm that this PR is working correctly in production. The Java version is 11 here instead of 8 https://sourcegraph.com/maven/com.spotify/dns/-/blob/lsif-java.json

**Pull request merged**